File: Algorithms/Misc/ManacherPalindrome.js
"use strict";

/*
 * Based on: https://articles.leetcode.com/longest-palindromic-substring-part-ii/
 */

let process = (s) => {

    let altChars = ['$', '#'];

    for (let char of s) {
        altChars.push(char);
        altChars.push('#');
    }

    altChars.push('@');

    return altChars;

}

let runManacher = (altChars) => {

    let palinTable = new Array(altChars.length);
    palinTable.fill(0);

    let center = 0,
        right = 0;

    for (let i = 1; i < altChars.length - 1; i++) {

        let opposite = center * 2 - i;

        if (right > i) {
            palinTable[i] = Math.min(right - i, palinTable[opposite])
        }

        while (altChars[i + (1 + palinTable[i])] == altChars[i - (1 + palinTable[i])]) {
            palinTable[i]++;
        }

        if (1 + palinTable[i] > right) {
            center = i;
            right = i + palinTable[i];
        }

    }

    return palinTable;

}

let longestPalindrome = (s) => {

    let altChars = process(s);
    let palinTable = runManacher(altChars);

    let longest = 0,
        center = 0;

    for (let i = 0; i < palinTable.length - 1; i++) {

        if (palinTable[i] > longest) {
            longest = palinTable[i];
            center = i;
        }

    }

    return s.substring((center - 1 - longest) / 2, (center - 1 + longest) / 2);

}

console.log(longestPalindrome("blahblahblahracecarblah"));

File: Algorithms/Misc/ManacherPalindrome.java
/*
 * Based on: https://articles.leetcode.com/longest-palindromic-substring-part-ii/
 */

public class ManacherPalindrome {

    private static char[] process(String s) {

        char[] altChars = new char[s.length() * 2 + 3];

        altChars[0] = '$';
        altChars[s.length() * 2 + 1] = '#';
        altChars[s.length() * 2 + 2] = '@';

        for (int i = 0; i < s.length(); i++) {

            altChars[i * 2 + 1] = '#';
            altChars[i * 2 + 2] = s.charAt(i);

        }

        return altChars;

    }

    private static int[] runManacher(char[] altChars) {

        int[] palinTable = new int[altChars.length];

        int center = 0, right = 0;

        for (int i = 1; i < altChars.length - 1; i++) {

            int opposite = center * 2 - i;

            if (right > i) {
                palinTable[i] = Math.min(right - i, palinTable[opposite]);
            }

            while (altChars[i + (1 + palinTable[i])] == altChars[i - (1 + palinTable[i])]) {
                palinTable[i]++;
            }

            if (i + palinTable[i] > right) {
                center = i;
                right = i + palinTable[i];
            }

        }

        return palinTable;

    }

    public static String longestPalindrome(String s) {

        char[] altChars = process(s);
        int[] palinTable = runManacher(altChars);

        int longest = 0, center = 0;

        for (int i = 0; i < palinTable.length - 1; i++) {

            if (palinTable[i] > longest) {
                longest = palinTable[i];
                center = i;
            }

        }

        return s.substring((center - 1 - longest) / 2, (center - 1 + longest) / 2);

    }

    public static void main(String[] args) {

        System.out.println(longestPalindrome("blahblahblahracecarblah"));

    }

}
